internal/webapp/oauth/oauth_localhost.go

```diff
@@ -0,0 +1,293 @@
+	}
+	if err := json.Unmarshal(body, &accountResponse); err != nil {
+		return "", fmt.Errorf("failed to parse account response: %w", err)
+	}
+
+	if accountResponse.Account.Team.UUID == "" {
+		return "", fmt.Errorf("team UUID not found in account response")
+	}
+
+	return accountResponse.Account.Team.UUID, nil
+}
+
+// --- Secret Storage ---
+
+// saveTokenToKeyring securely stores the token in the system's native keychain.
+func saveTokenToKeyring(teamUUID, token string) error {
+	err := keyring.Set(keyringService, teamUUID, token)
+	if err != nil {
+		return fmt.Errorf("failed to save token to keyring: %w", err)
+	}
+	err = keyring.Set(keyringService, keyringLastUsedTeam, teamUUID)
+	if err != nil {
+		return fmt.Errorf("failed to save token to keyring: %w", err)
+	}
+	slog.Info(fmt.Sprintf("Successfully saved token for team %s to the system keyring.", teamUUID))
+	return nil
+}
+
+// GetLastUsedTeamUUID retrieves the last saved team UUID from the keyring.
+func GetLastUsedTeamUUID() (string, error) {
+	secret, err := keyring.Get(keyringService, keyringLastUsedTeam)
+	if err != nil {
+		return "", err
+	}
+	return secret, nil
+}
+
+// GetTokenFromKeyring retrieves a token from the keyring for a given team UUID.
+func GetTokenFromKeyring(teamUUID string) (string, error) {
+	secret, err := keyring.Get(keyringService, teamUUID)
+	if err != nil {
+		return "", err
+	}
+	return secret, nil
+}
+
+// --- Handlers ---
+
+// callbackHandler handles the OAuth2 callback.
+// It uses a two-step process for the implicit grant flow.
+func callbackHandler(tokenChan chan string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// Step 2: Check if the access_token is now in the query parameters
+		// after being redirected by our own JavaScript.
+		token := r.URL.Query().Get("access_token")
+
+		// If the token is NOT in the query string, this is the first hit from DigitalOcean.
+		// The token is in the URL fragment ('#'), which the server cannot see.
+		if token == "" {
+			// Step 1: Serve an HTML page with JavaScript.
+			// This script will read the token from the fragment and reload the page,
+			// moving the token into the query string ('?') for the server to read.
+			w.Header().Set("Content-Type", "text/html")
+			fmt.Fprint(w, `
+<!DOCTYPE html>
+<html>
+<head>
+		<title>Authorizing...</title>
+		<script>
+				window.addEventListener('DOMContentLoaded', () => {
+						const url = new URL(window.location.href);
+						// Check if the URL has a hash component.
+						const hash = url.hash.slice(1);
+
+						// Only process if there is a hash.
+						if (hash) {
+								const searchParams = new URLSearchParams(hash);
+								// If the access_token exists in the hash, move it to the query string.
+								if (searchParams.has("access_token")) {
+										url.search = searchParams.toString();
+										url.hash = "";
+										// Replace the URL to reload the page with the token in the query string.
```

Comment:
FYI, this exposes the API token in a plaintext (HTTP) request, which security previously rejected when I attempted to follow this same approach in `doctl` a few years ago.

They might be okay with it now if you switch from `localhost` to the loopback `127.0.0.1` given it's ["potentially trustworthy"](https://www.w3.org/TR/secure-contexts/#is-origin-trustworthy) (see the 4th item in that list).

Reply:
...Oh wait you are the security team, I opened this PR from a link and didn't check the author 😅